{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Great, now that we discussed a little let's continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given that the current approach utilized by the authors lacks reproducibility, we will explore an alternative method by leveraging nf-core pipelines for data analysis.\n",
    "\n",
    "Please explain, how we will achieve reproducibility for the course  with this approach.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Nf-core pipelines achieve reproducibility based on extensive documentation, stable releases, open-source code, and packaged software."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You have successfully downloaded 2 of the fastq files we will use in our study.\n",
    "\n",
    "What is the next step if we want to first have a count table and check the quality of our fastq files? What is the pipeline called to do so?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- We first do quality control of the FASTQ files using FastQC\n",
    "- We can use nf-core/rnaseq for the RNA-seq data processing. This allows reproducibility."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze the 2 files using an nf-core pipeline.\n",
    "\n",
    "What does this pipeline do?\n",
    "\n",
    "Which are the main tools that will be used in the pipeline?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What does this pipeline do?\n",
    "- The pipeline nf-core/rnaseq is a pipeline that processes RNA-seq data starting with the raw reads. \n",
    "- The pipeline can be \n",
    "\n",
    "Which are the main tools that will be used in the pipeline?\n",
    "1. Pre-processing: The tool does quality control of the FASTQ files (tool: FastQC). Also it uses trimming to remove low-quality bases and adapter sequences from the reads (tool: Trim Galore!)\n",
    "2. Genome alignment & quantification: It then counts the number of reads mapped to each gene to generate a gene count table. (Alignment tool: STAR or HISAT2)\n",
    "3. Pseudo-alignment & quantification (tools: Salmon, Kallisto)\n",
    "4. Post-Processing (SAMtools)\n",
    "5. Final QC: Afterwards, a report is generated that combines all quality metrics to assess the success of the analysis. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As all other nf-core pipelines, the chosen pipeline takes in a samplesheet as input.\n",
    "\n",
    "Use Python and pandas to create the samplesheet for your 2 samples. Feel free to make use of the table you created earlier today.\n",
    "\n",
    "Choose your sample names wisely, they must be the connection of the results to the metadata. If you can't find the sample in the metadata later, the analysis was useless."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    sample                                            fastq_1  \\\n",
      "0  SRX19144486_SRR23195516  /home/tabea/ComputationalWorkflows/Tag2/SRA_fi...   \n",
      "1  SRX19144488_SRR23195511  /home/tabea/ComputationalWorkflows/Tag2/SRA_fi...   \n",
      "\n",
      "                                             fastq_2 strandedness  \n",
      "0  /home/tabea/ComputationalWorkflows/Tag2/SRA_fi...         auto  \n",
      "1  /home/tabea/ComputationalWorkflows/Tag2/SRA_fi...         auto  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Creating the samplesheet with the necessary columns: 'sample', 'fastq_1', 'fastq_2'\n",
    "samplesheet_dict = {\"sample\": [\"SRX19144486_SRR23195516\", \"SRX19144488_SRR23195511\"],\n",
    "               \"fastq_1\": [\"/home/tabea/ComputationalWorkflows/Tag2/SRA_files/fastq/SRX19144486_SRR23195516_1.fastq.gz\", \"/home/tabea/ComputationalWorkflows/Tag2/SRA_files/fastq/SRX19144486_SRR23195516_2.fastq.gz\"],\n",
    "               \"fastq_2\": [\"/home/tabea/ComputationalWorkflows/Tag2/SRA_files/fastq/SRX19144488_SRR23195511_1.fastq.gz\", \"/home/tabea/ComputationalWorkflows/Tag2/SRA_files/fastq/SRX19144488_SRR23195511_2.fastq.gz\"],\n",
    "               \"strandedness\": [\"auto\", \"auto\"]}\n",
    "\n",
    "\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "samplesheet = pd.DataFrame(samplesheet_dict)\n",
    "\n",
    "# Save the samplesheet as a CSV file for nf-core pipeline\n",
    "samplesheet.to_csv('samplesheet.csv', index=False)\n",
    "\n",
    "# Display the samplesheet\n",
    "print(samplesheet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mNextflow 24.04.4 is available - Please consider updating your version to it\u001b[m\n",
      "N E X T F L O W  ~  version 23.10.0\n",
      "NOTE: Your local project version looks outdated - a different revision is available in the remote repository [33df0c05ef]\n",
      "Launching `https://github.com/nf-core/rnaseq` [festering_varahamihira] DSL2 - revision: 4053b2ec17 [master]\n",
      "\u001b[33mWARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\u001b[39m\u001b[K\n",
      "\n",
      "\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "                                        \u001b[0;32m,--.\u001b[0;30m/\u001b[0;32m,-.\u001b[0m\n",
      "\u001b[0;34m        ___     __   __   __   ___     \u001b[0;32m/,-._.--~'\u001b[0m\n",
      "\u001b[0;34m  |\\ | |__  __ /  ` /  \\ |__) |__         \u001b[0;33m}  {\u001b[0m\n",
      "\u001b[0;34m  | \\| |       \\__, \\__/ |  \\ |___     \u001b[0;32m\\`-._,-`-,\u001b[0m\n",
      "                                        \u001b[0;32m`._,._,'\u001b[0m\n",
      "\u001b[0;35m  nf-core/rnaseq v3.15.1-g4053b2e\u001b[0m\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "\u001b[1mCore Nextflow options\u001b[0m\n",
      "  \u001b[0;34mrevision        : \u001b[0;32mmaster\u001b[0m\n",
      "  \u001b[0;34mrunName         : \u001b[0;32mfestering_varahamihira\u001b[0m\n",
      "  \u001b[0;34mcontainerEngine : \u001b[0;32mdocker\u001b[0m\n",
      "  \u001b[0;34mlaunchDir       : \u001b[0;32m/home/tabea/ComputationalWorkflows/Tag2/day2/day2\u001b[0m\n",
      "  \u001b[0;34mworkDir         : \u001b[0;32m/home/tabea/ComputationalWorkflows/Tag2/day2/day2/work\u001b[0m\n",
      "  \u001b[0;34mprojectDir      : \u001b[0;32m/home/tabea/.nextflow/assets/nf-core/rnaseq\u001b[0m\n",
      "  \u001b[0;34muserName        : \u001b[0;32mtabea\u001b[0m\n",
      "  \u001b[0;34mprofile         : \u001b[0;32mdocker\u001b[0m\n",
      "  \u001b[0;34mconfigFiles     : \u001b[0;32m\u001b[0m\n",
      "\n",
      "\u001b[1mInput/output options\u001b[0m\n",
      "  \u001b[0;34minput           : \u001b[0;32m/home/tabea/ComputationalWorkflows/Tag2/day2/day2/samplesheet.csv\u001b[0m\n",
      "  \u001b[0;34moutdir          : \u001b[0;32m/home/tabea/ComputationalWorkflows/Tag2/day2/day2/rnaseq-output\u001b[0m\n",
      "\n",
      "\u001b[1mReference genome options\u001b[0m\n",
      "  \u001b[0;34mfasta           : \u001b[0;32m/home/tabea/ComputationalWorkflows/Tag2/day2/day2/GCF_000001635.27_GRCm39_genomic.fna\u001b[0m\n",
      "  \u001b[0;34mgtf             : \u001b[0;32m/home/tabea/ComputationalWorkflows/Tag2/day2/day2/genomic.gtf\u001b[0m\n",
      "\n",
      "\u001b[1mAlignment options\u001b[0m\n",
      "  \u001b[0;34mmin_mapped_reads: \u001b[0;32m5\u001b[0m\n",
      "\n",
      "\u001b[1mMax job request options\u001b[0m\n",
      "  \u001b[0;34mmax_memory      : \u001b[0;32m7.GB\u001b[0m\n",
      "\n",
      "!! Only displaying parameters that differ from the pipeline defaults !!\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "If you use nf-core/rnaseq for your analysis please cite:\n",
      "\n",
      "* The pipeline\n",
      "  https://doi.org/10.5281/zenodo.1400710\n",
      "\n",
      "* The nf-core framework\n",
      "  https://doi.org/10.1038/s41587-020-0439-x\n",
      "\n",
      "* Software dependencies\n",
      "  https://github.com/nf-core/rnaseq/blob/master/CITATIONS.md\n",
      "-\u001b[2m----------------------------------------------------\u001b[0m-\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "\u001b[6A\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "\u001b[15A\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DESEQ2... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:STRING... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:SUBREA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQ... -\u001b[K\n",
      "\u001b[34A\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DESEQ2... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:STRING... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:SUBREA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQ... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUALIM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DUPRADAR  -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "\u001b[49A\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DESEQ2... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:STRING... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:SUBREA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQ... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUALIM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DUPRADAR  -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQC   -\u001b[K\n",
      "\u001b[50A\n",
      "executor >  local (2)\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... [  0%] 0 of 1\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... [  0%] 0 of 1\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[23/e8a5f4] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... [  0%] 0 of 2\u001b[K\n",
      "[21/dc9d83] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... [  0%] 0 of 2\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DESEQ2... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:STRING... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:SUBREA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQ... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUALIM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DUPRADAR  -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQC   -\u001b[K\n",
      "\u001b[51A\n",
      "-\u001b[0;35m[nf-core/rnaseq]\u001b[0;31m Pipeline completed with errors\u001b[0m-\n",
      "executor >  local (2)\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... [  0%] 0 of 1\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... [  0%] 0 of 1\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:PREPARE_GENOM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[23/e8a5f4] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... [  0%] 0 of 2\u001b[K\n",
      "[21/dc9d83] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... [  0%] 0 of 2\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:FASTQ_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:ALIGN_... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUANTI... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DESEQ2... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_MA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:STRING... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:SUBREA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQ... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDTOO... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BEDGRA... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:QUALIM... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:DUPRADAR  -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:BAM_RS... -\u001b[K\n",
      "[-        ] process > NFCORE_RNASEQ:RNASEQ:MULTIQC   -\u001b[K\n",
      "\u001b[31mERROR ~ Pipeline failed. Please refer to troubleshooting docs: https://nf-co.re/docs/usage/troubleshooting\u001b[K\n",
      "\u001b[K\n",
      " -- Check '.nextflow.log' file for details\u001b[39m\u001b[K\n",
      "WARN: Killing running tasks (2)\n",
      "\n",
      "ERROR ~ Unexpected error [NullPointerException]\n",
      "\n",
      " -- Check script '/home/tabea/.nextflow/assets/nf-core/rnaseq/./workflows/rnaseq/../../subworkflows/nf-core/utils_nfcore_pipeline/main.nf' at line: 105 or see '.nextflow.log' file for more details\n"
     ]
    }
   ],
   "source": [
    "# post here the command you used to run nf-core/rnaseq\n",
    "\n",
    "! nextflow run nf-core/rnaseq --input /home/tabea/ComputationalWorkflows/Tag2/day2/day2/samplesheet.csv --outdir /home/tabea/ComputationalWorkflows/Tag2/day2/day2/rnaseq-output --gtf /home/tabea/ComputationalWorkflows/Tag2/day2/day2/genomic.gtf --fasta /home/tabea/ComputationalWorkflows/Tag2/day2/day2/GCF_000001635.27_GRCm39_genomic.fna -profile docker"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Explain all the parameters you set and why you set them in this way."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- I had to change the max_memory in the configuration file of nextflow. \n",
    "- Also, I had to define the input with --input, the output direction with --outdir, .gtf file of the reference genome with --gtf, reference genome using --fasta and the profile, in this case -profile docker"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Browsing the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How did the pipeline perform?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- It took long to run through. Also, it needed lots of memory, which is why some configuration parameters had to be adapted to fit my laptop."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Explain the quality control steps. Are you happy with the quality and why. If not, why not.\n",
    "Please give additional information on : \n",
    "- ribosomal rRNA\n",
    "- Duplication\n",
    "- GC content\n",
    "\n",
    "What are the possible steps that could lead to poorer results?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Explain the quality control steps.\n",
    "- Quality control steps include evaluating the quality of raw reads using FastQC. The resulting FastQC report includes basic statistics, per base sequence quality, per tile sequence quality, per sequence quality scores, per base sequence content, per sequence GC content, per base N content, sequence length distribution, sequence duplication levels, overrepresented sequences, and adapter content.\n",
    "\n",
    "Are you happy with the quality and why. If not, why not.\n",
    "- Overall, the results are improvable, but we can work with the data. The per base sequence quality, per sequence quality score , per base N content, and the sequence length distribution is good in all reports. However, the adapter content in all samples is quite high, we often deal with sequence duplication levels, per base sequence content, per sequence GC content, and also with overrepresented sequences. \n",
    "\n",
    "Explanation of each summary statistic:\n",
    "1. Per base sequence quality: A sequence quality score of 30 or above in the the results of sequencing quality per base is considered good. For Sham_oxy_1_1.gz, ... the average sequence quality per base is within this range. \n",
    "\n",
    "2. Per tile sequence quality:\n",
    "The quality of each tile sequenced can be determined through a color scale. A tile is considered good, when it consists of cooler colors indicating higher quality. Hotter colors indicate lower quality. \n",
    "\n",
    "3. Per sequence quality scores: 90% of the reads with a quality score of 35 or above are considered to have good quality. If we have lots of low quality scores, there could be a problem with the sequencing run itself. \n",
    "\n",
    "4. Per base sequence content: In a high-quality sequencing sample, the four lines indicating the base proportions at each position should be nearly parallel and closely aligned\n",
    "\n",
    "5. Per sequence GC content: The GC content distribution is analyzed by comparing the observed distribution to a reference normal distribution, helping identify any biases or contamination in the sequencing library; deviations greater than 15% are flagged as warnings, while those exceeding 30% are considered errors.\n",
    "\n",
    "6. Per base N content: The N ratio per base indicates the percentage of indeterminate bases in sequencing reads, and if it exceeds 5%, it serves as a warning of potential sequencing issues, while values over 20% are considered errors, warranting investigation to maintain data quality.\n",
    "\n",
    "7. Sequence Length Distribution: A uniform sequence length distribution indicates all sequences in the library have the same length, which is typical for platforms like Illumina; non-uniform lengths may require trimming or specialized tools for downstream analyses, and reads of length 0 signal errors that must be addressed.\n",
    "\n",
    "8. Sequence Duplication Levels: Repeated sequences measure duplication or bias in sequencing data; while some duplication is expected with high sequencing depth, over 20% duplication may be a warning, and exceeding 50% signals a significant issue that must be corrected to ensure accurate downstream analysis.\n",
    "\n",
    "9. Overrepresented Sequences: An over-represented sequence is one that occurs frequently in sequencing reads, with a threshold of 0.1% for warning and 1% for an error, assessed on the first 200,000 reads to detect potential contaminants or biases.\n",
    "\n",
    "10. Adapter Content: Adapter sequences are short DNA fragments attached to the ends of DNA before sequencing, and their presence is tracked as a cumulative percentage to assess how frequently they appear at different positions in the sequenced DNA fragments.\n",
    "\n",
    "What are the possible steps that could lead to poorer results?\n",
    "- Sequencing errors (e.g., poor quality reads at the end of the run) can lead to poorer results.\n",
    "- Also, adapter sequences that are not properly removed from reads can further influence the results.\n",
    "- Furthermore, low-quality material in the beginning caused by degration, insufficient removal of rRNA etc. can lead to poorer results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Would you exclude any samples? If yes, which and why?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In theory, the goal is to drop as little samples as possible, especially when dealing with a small amount of samples in the beginning. However, achieving only high quality data isn't as simple as it sound. Typically, one looks at the FastQC results to decide whether the sample should be left inside or dropped from the dataset.\n",
    "- In this case, I would try to keep as many files as possible because of the small amount of samples. Sham_oxy_1_1.z seems to have the worst quality report, not achieving good quality in per sequence GC content, sequence duplication levels, overrepresented sequences, and adapter content. Also I would drop Sham_oxy_1_2.gz as it needs improvement in per base sequence content, per sequence GC content, sequence duplication levels, overrepresented sequences, and seq with adapter content."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What would you now do to continue the experiment? What are the scientists trying to figure out? Which packages on R or python would you use?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To continue the experiment, I would proceed with differential abundance analysis to determine which features (such as genes, species, or proteins) show significant differences in abundance. For this differential analysis, using the nf-core/differentialabundance pipeline is appropriate."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextflow",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
